[Spark-9028] [ML] Add CountVectorizer as an estimator to generate CountVectorizerModel

hhbyyh opened the pull request:
jira: https://issues.apache.org/jira/browse/SPARK-9028 

Add an estimator for CountVectorizerModel. The estimator will extract a vocabulary from document collections according to the term frequency.

I changed the meaning of minCount as a filter across the corpus. This aligns with Word2Vec and the similar parameter in SKlearn.

hhbyyh committed:
count vectorizer estimator

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into cvEstimator

hhbyyh committed:
add more ut for estimator

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37187/consoleFull) for   PR 7388 at commit [`93e1ad4`](https://github.com/apache/spark/commit/93e1ad494b837a4b82b95ac1001e1b1adafe4df1).

SparkQA commented:
[Test build #37187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37187/console) for   PR 7388 at commit [`93e1ad4`](https://github.com/apache/spark/commit/93e1ad494b837a4b82b95ac1001e1b1adafe4df1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CountVectorizer(override val uid: String)`
  * `class CountVectorizerModel(override val uid: String, val vocabulary: Array[String])`
  * `case class Least(children: Expression*) extends Expression `
  * `case class Greatest(children: Expression*) extends Expression `

AmplabJenkins commented:
Merged build finished. Test PASSed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,135 @@ package org.apache.spark.ml.feature
  * :: Experimental ::
- * Converts a text document to a sparse vector of token counts.
- * @param vocabulary An Array over terms. Only the terms in the vocabulary will be counted.
+ * Extracts a vocabulary from document collections and generates a CountVectorizerModel.
```
"[[CountVectorizerModel]]"

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,135 @@ package org.apache.spark.ml.feature
+  override def fit(dataset: DataFrame): CountVectorizerModel = {
+    transformSchema(dataset.schema, logging = true)
+    val input = dataset.select($(inputCol)).map(_.getAs[Seq[String]](0))
+    val min_count = $(minCount)
```
No need for this, just use `$(minCount)` in the code. Also, in general prefer `camelCase` over `snake_case`

hhbyyh replied:
no performance difference?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,135 @@ package org.apache.spark.ml.feature
+    transformSchema(dataset.schema, logging = true)
+    val input = dataset.select($(inputCol)).map(_.getAs[Seq[String]](0))
+    val min_count = $(minCount)
+    val vocab_size = $(vocabSize)
```
ditto to L98

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,135 @@ package org.apache.spark.ml.feature
+      tmpSortedWC.map(_._1)
+    }
+
+    require(vocab.length > 0, "The vocabulary size should be > 0. Adjust minCount as necessary.")
```
Should this throw an exception or just log an error? I'm not too sure...

feynmanliang replied:
If we do decide to throw and exception, we should add a test for it

hhbyyh replied:
Thanks, I'll add an ut.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,135 @@ package org.apache.spark.ml.feature
+    val vectorizer = udf { (document: Seq[String]) =>
       val termCounts = mutable.HashMap.empty[Int, Double]
       document.foreach { term =>
         dict.get(term) match {
```
Am I just missing the `case Some(...)`, or is it not there?

feynmanliang commented:
That's all for now!

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into cvEstimator

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into cvEstimator

hhbyyh committed:
minor fix

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38415/consoleFull) for   PR 7388 at commit [`589e93d`](https://github.com/apache/spark/commit/589e93d3267b8a3f9ef1f1aba3aec22463dddfa6).

SparkQA commented:
[Test build #38415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38415/console) for   PR 7388 at commit [`589e93d`](https://github.com/apache/spark/commit/589e93d3267b8a3f9ef1f1aba3aec22463dddfa6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class CountVectorizer(override val uid: String)`
  * `class CountVectorizerModel(override val uid: String, val vocabulary: Array[String])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

hhbyyh commented:
@feynmanliang  Thanks for helping review. Sent an update

jkbradley commented:
I'll make a pass now.  Sorry for the long delay!

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,133 @@ package org.apache.spark.ml.feature
+private[feature] trait CountVectorizerParams extends Params with HasInputCol with HasOutputCol {
+
+  /**
+   * size of the vocabulary.
```
"size" --> "Max size"

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,133 @@ package org.apache.spark.ml.feature
+  }
+
+  override def validateParams(): Unit = {
+    require($(vocabSize) > 0, s"The vocabulary size (${$(vocabSize)}) must be above 0.")
```
This should be checked in the Param (via the isValid argument).

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,133 @@ package org.apache.spark.ml.feature
-@Experimental
-class CountVectorizerModel (override val uid: String, val vocabulary: Array[String])
-  extends UnaryTransformer[Seq[String], Vector, CountVectorizerModel] {
+class CountVectorizer(override val uid: String)
```
Experimental

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizerModel.scala:
```diff
@@ -19,45 +19,133 @@ package org.apache.spark.ml.feature
+    val wordCounts: RDD[(String, Long)] = input
+      .flatMap { case (tokens) => tokens.map(_ -> 1L) }
+      .reduceByKey(_ + _)
+      .filter(_._2 >= $(minCount))
```
This and other uses of Params should be made into local copies before RDD operations, like you had originally.  There is a minor cost to the lookup, and it's good practice to prevent the whole class from being serialized (though it does not matter for this case).

hhbyyh replied:
Yes, I thought so... There're some other transformers (like tokenizer) those uses $ in their transform. Maybe we should change them?

jkbradley replied:
That sounds good, but it's not high priority (so 1.6, not for 1.5)

jkbradley commented:
Why did you remove minTermFreq?  I think it's a useful way to make things sparser.

Also, could you change "minCount" to instead be "minDocFreq" (which has slightly different semantics, operating on doc frequency instead of word counts)?  That will match the semantics in sklearn ("min_df").  If it's easier, I'm also OK with not providing this option for the initial version, and adding it for the next release instead.

jkbradley commented:
Those are my initial comments.  I'll check back!

jkbradley commented:
One more comment: When you move a file, it helps to use git mv so that Github understands it's the same file (for the diff).  No big deal here though.

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into cvEstimator

hhbyyh committed:
Merge branch 'cvEstimator' of https://github.com/hhbyyh/spark into cvEstimator

hhbyyh commented:
Thanks for taking the time review. 

About minTermFreq, I renamed it to minCount since I thought users may misunderstand minTermFreq as a double value (like 0.01, 0.05). Anyway, minDocFreq is a great idea. I'll use it to replace minTermFreq.

The hint about git mv is noted. Sorry for the extra effort during review.

jkbradley commented:
I actually agree with you about "freq" being misleading, but we decided to stick with it since it's used to mean "count" in many libraries.  (This discussion was on some JIRA or PR, but I forget which one..)

No problem about git mv  : )

jkbradley commented:
Actually I think I have time to help a little now.  I'll send some updates soon.

hhbyyh commented:
Please. Thanks a lot.

jkbradley committed:
Updates:
* Renamed "minCount" to "minTokenCount"
* Added "minTermFreq" back, including unit test
* Moved all Params to include in both Estimator and Model so that they can be viewed in either.

hhbyyh committed:
Merge pull request #3 from jkbradley/countvec

Updated CountVectorizer

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

hhbyyh commented:
Thanks @jkbradley for the update. Copy the comment of Joseph from https://github.com/hhbyyh/spark/pull/3 to here:

Updates:

Renamed "minCount" to "minTokenCount"
Added "minTermFreq" back, including unit test
Moved all Params to include in both Estimator and Model so that they can be viewed in either.
New thoughts:

Previously, we had decided to call "minTermFreq" a "frequency" even though it is a count. That was to follow other things such as sklearn and "TF-IDF."
Now, I would like to rename it to "minTermCount." Yuhao got me thinking along these lines, and it actually matches other libraries to some extent:
sklearn calls it "min_tf," but it accepts both integers (counts) and doubles (fractions or frequencies). Since we cannot take both the way Params are set up, the best we could do is have a minTermCount Param and a minTermFreq Param. Frequency might actually be a better way to specify this threshold.
TF-IDF arguably uses fractional frequencies (not counts)...depending on whether "TF-IDF" refers to the value or the normalizer.
Proposal: I propose we do 1 of the following:

(A) Rename "minTermFreq" to "minTermCount"
(B) Change "minTermFreq" to take a Double fraction
(C) Remove "minTermFreq" for now pending further discussion
CC: @hhbyyh @mengxr What do you think?

@hhbyyh Feel free to merge, take parts, or modify as you see fit. It may be easiest if you go ahead and merge/modify so that @mengxr can view your full PR.

Thanks!

SparkQA commented:
[Test build #40881 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40881/consoleFull) for   PR 7388 at commit [`7d4743a`](https://github.com/apache/spark/commit/7d4743a395f924ca82af8a136e33d69f906c40ff).

SparkQA commented:
[Test build #40881 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40881/console) for   PR 7388 at commit [`7d4743a`](https://github.com/apache/spark/commit/7d4743a395f924ca82af8a136e33d69f906c40ff).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,196 @@
+   * Default: 1
+   * @group param
+   */
+  val minTokenCount: IntParam = new IntParam(this, "minTokenCount",
```
Is this parameter more useful than `minDocFreq`/`maxDocFreq`? If there exists a single doc that contains a spam word for many times, this would include that word in the vocabulary.

jkbradley replied:
I'm fine with removing that.

hhbyyh replied:
Then we can have minDocFreq only for this release.

hhbyyh replied:
I'll try to replace it with minDocFreq

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,196 @@
+   * Default: 1
+   * @group param
+   */
+  val minTermFreq: IntParam = new IntParam(this, "minTermFreq",
```
Should mention that this doesn't affect fitting.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,196 @@
+   * CountVectorizer will build a vocabulary that only considers the top
+   * vocabSize terms ordered by term frequency across the corpus.
+   *
+   * Default: 10000
```
Any reference to this default value? I like a default value that works for most cases than not putting a limit by default. But I don't know what would be a good default value. It might be hard to change this default in the future.

jkbradley replied:
sklearn uses inf, but do we really want to encourage that?

mengxr replied:
No, I don't think `inf` is a good default. Just wonder what would be a reasonable default value here.

jkbradley replied:
We could set it to match HashingTF (1 << 18)

hhbyyh replied:
I'll set it to 1 << 18 for now. This should cover the most common English words.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/feature/CountVectorizerSuite.scala:
```diff
@@ -0,0 +1,137 @@
+      .setMinTermFreq(3)
+    val output = cv.transform(df).collect()
+    output.foreach { p =>
+      val features = p.getAs[Vector]("features")
```
Maybe simpler to do `cv.transform(df).select("features", "expected").collect()` then `foreach { case Row(features: Vector, expected: Vector) => ...`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,196 @@
+  def setMinTermFreq(value: Int): this.type = set(minTermFreq, value)
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    val dict = vocabulary.zipWithIndex.toMap
```
minor: `dict` should be a private broadcast variable, because `transform` might be called multiple times in an ML pipeline.

jkbradley replied:
Hm, this is an issue multiple places in MLlib.  I'll make a JIRA for it...

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,196 @@
+   * @group param
+   */
+  val vocabSize: IntParam =
+    new IntParam(this, "vocabSize", "size of the vocabulary", ParamValidators.gt(0))
```
`size` -> `max size`

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,196 @@
+  def setMinTokenCount(value: Int): this.type = set(minTokenCount, value)
+
+  /** @group setParam */
+  def setMinTermFreq(value: Int): this.type = set(minTermFreq, value)
```
I removed the interface since minTermFreq is not used during fitting.

hhbyyh committed:
replace minTokenCount with minDocFreq

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #40940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40940/consoleFull) for   PR 7388 at commit [`17b3009`](https://github.com/apache/spark/commit/17b30097d0a553fc74cd0aef03ff8ca1ead4dfb7).

SparkQA commented:
[Test build #40940 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40940/console) for   PR 7388 at commit [`17b3009`](https://github.com/apache/spark/commit/17b30097d0a553fc74cd0aef03ff8ca1ead4dfb7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/40940/
Test PASSed.

jkbradley commented:
Talked with @mengxr and have slightly different plan for freq vs. count.  We'll switch to sklearn's method and create new params call minTF and minDF as needed.  I'll send a PR to update this one accordingly.  I'll also include:
* setMinTermFreq in the Estimator so that users can specify all parameters before fitting and have them passed to the model.
* private var for broadcast

ETA 30 min

jkbradley commented:
ETA 9 more min

jkbradley commented:
waiting for tests to run before I send it...

jkbradley committed:
renamed docFreq to DF, termFreq to TF, and added fractional support.  save broadcast as private var

jkbradley commented:
OK sent

hhbyyh committed:
Merge pull request #4 from jkbradley/cntvec-update

docFreq, termFreq, broadcast update

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41070 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41070/consoleFull) for   PR 7388 at commit [`a5a8532`](https://github.com/apache/spark/commit/a5a8532a5705499447b429768cf4dbb1484eee6a).

hhbyyh commented:
Merged commit of @jkbradley from https://github.com/hhbyyh/spark/pull/4. Thanks Joseph, the code looks much better.

jkbradley commented:
Btw, I was wrong about "git mv".  Github still doesn't do the diff properly if you make any changes to the file after moving it.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,235 @@
+    val minDf: Long = if ($(minDF) >= 1.0) {
+      $(minDF).toLong
+    } else {
+      math.ceil($(minDF) * input.cache().count()).toLong
```
I don't really like this extra cache, but I'd guess it's better than not caching.

hhbyyh replied:
I think it's good...

SparkQA commented:
[Test build #41070 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41070/console) for   PR 7388 at commit [`a5a8532`](https://github.com/apache/spark/commit/a5a8532a5705499447b429768cf4dbb1484eee6a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41070/
Test PASSed.

jkbradley commented:
If this looks good to @hhbyyh , I'll wait for a final API check by @mengxr before merging this

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,235 @@
+    val vocSize = $(vocabSize)
+    val input = dataset.select($(inputCol)).map(_.getAs[Seq[String]](0))
+    val minDf: Long = if ($(minDF) >= 1.0) {
+      $(minDF).toLong
```
`math.ceil($(minDF)).toLong`

hhbyyh replied:
The update will "eat" the unexpected input of minDF from users, like 3.1 or 2.5. Will it be misleading?

mengxr replied:
If a user set `minDF` to `5.5`, the expected `minDF` should be `6`. I don't think the user would expect anything else. Btw, a simpler implementation is to keep `minDf` as a `Double`:

~~~scala
val minDf = if ($(minDF) >= 1.0) {
  $(minDF)
} else {
  $(minDF) * input.cache().count()
}
~~~

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/feature/CountVectorizer.scala:
```diff
@@ -0,0 +1,235 @@
+      tokens.foreach { w =>
+        wc.changeValue(w, 1L, _ + 1L)
+      }
+      wc.map { case (word, count) => (word, (count, 1)) }
```
* `1` -> `1L`
* Is it useful to keep the total count? We can order terms by DF. It could simplify the implementation since we only need distinct: `tokens.distinct.flatMap((_, 1L))`

jkbradley replied:
sklearn seems to use the total count: [https://github.com/scikit-learn/scikit-learn/blob/a95203b249c1cf392f86d001ad999e29b2392739/sklearn/feature_extraction/text.py#L704]
I could see arguments for either, but I'd lean towards total count since you could have a corpus where a small number of docs (small DF) have many instances of some term (medium/high token count), in which case removing those terms could remove significant chunks of the docs.

hhbyyh replied:
Just some shallow thought, Ordering terms by total count seems more intuitive and is consistent with sklearn, while ordering by DF might be of a little surprise to users. Please comment freely.

mengxr replied:
Okay, let's keep it as is.

hhbyyh committed:
Merge remote-tracking branch 'upstream/master' into cvEstimator

hhbyyh committed:
Merge branch 'cvEstimator' of https://github.com/hhbyyh/spark into cvEstimator

hhbyyh committed:
use minDF as Double

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

hhbyyh commented:
@mengxr @jkbradley Sent an update addressing the comment.

SparkQA commented:
[Test build #41146 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41146/consoleFull) for   PR 7388 at commit [`a370816`](https://github.com/apache/spark/commit/a37081636388546f7c84e429d38f206b014cc88e).

jkbradley commented:
Modifying the LDA example sounds good to me.  It does mean using spark.ml classes in spark.mllib examples, but I don't really see a problem with that for examples.  (I would not want to do that in the main code though.)

SparkQA commented:
[Test build #41146 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41146/console) for   PR 7388 at commit [`a370816`](https://github.com/apache/spark/commit/a37081636388546f7c84e429d38f206b014cc88e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41146/
Test PASSed.

jkbradley commented:
I'll merge this with master and branch-1.5.  @hhbyyh Thanks a lot!
